monix-tail/shared/src/main/scala/monix/tail/internal/IterantRepeat.scala

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2014-2018 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.tail.internal
+
+import cats.effect.Sync
+import monix.execution.misc.NonFatal
+import monix.tail.Iterant
+import monix.tail.Iterant.{Halt, Last, Next, NextBatch, NextCursor, Suspend}
+import monix.tail.internal.IterantUtils.signalError
+import cats.syntax.all._
+
+private[tail] object IterantRepeat {
+  /** Implementation for `Iterant.repeat`. */
+  def apply[F[_], A, B](source: Iterant[F, A])(implicit F: Sync[F]): Iterant[F, A] = {
+
+    def loop(self: Iterant[F, A]): Iterant[F, A] =
+      try self match {
+        case Next(head, tail, stop) =>
+          Next[F, A](head, tail.map(loop), stop)
+        case NextCursor(cursor, rest, stop) =>
+          NextCursor[F, A](cursor, rest.map(loop), stop)
+        case NextBatch(gen, rest, stop) =>
+          NextBatch(gen, rest.map(loop), stop)
+        case Suspend(rest, stop) =>
+          Suspend[F, A](rest.map(loop), stop)
+        case Last(item) =>
+          Next[F, A](item, F.delay(loop(source)), F.unit)
+        case Halt(Some(ex)) =>
+          signalError(self, ex)
+        case Halt(None) =>
+          Suspend(F.delay(loop(source)), F.unit)
+      } catch {
+        case ex if NonFatal(ex) => signalError(source, ex)
+      }
+
+    source match {
+      // terminate if the source is empty
+      case empty @ Halt(_) => empty
```

Comment:
So this is good, however there's a catch.

If you'll take a look at [RepeatSourceObservable](https://github.com/monix/monix/blob/master/monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/RepeatSourceObservable.scala), it has an `isEmpty: Boolean` used to track if the source emitted anything or not.

And this check right here is not sufficient. For one because you could be dealing with a `Suspend(Halt(None))`.
`NextCursor` is problematic because you could have a `NextCursor().repeat`. Now if the user does that, then fine, he probably knows what he's doing, but with no more elements to repeat, we need to stop the loop.

So I'm thinking about this

1. keep an `isEmpty` state in that loop
2.  only as long as `isEmpty == true`, when receiving a `NextCursor` we need to check if it `hasNext`, such that we set `isEmpty` accordingly, because we don't want to repeat a stream made of one empty `NextCursor`
3. only as long as `isEmpty == true`, we need to unwrap `NextBatch` into `NextCursor` such that we can check if it `hasNext` and set `isEmpty` accordingly
4. because we end up evaluating cursors, this means that in this match right here we need to `Suspend` in case we have `NextBatch` or `NextCursor`

Reply:
@alexandru 
Ok, I've added handling cases like ```Suspend(Halt(None))``` and ```NextCursor(BatchCursor(), F(BatchCursor(1, 2, 3)))```. There is ugly nested `if` in ```case NextBatch(_, _, _)```